### DISCUSSION TITLE
Refaceting summary and resolution

### DISCUSSION THREAD

**Discussion by dkowollik** [ansys/pyprimemesh#870] [Q&A] [answered]
Hi, 
first of all, is CadRefaceting only available during import of models or is it possible to refacet afterwards? If this is possible, can someone please provide python snippets to do so?

Furthermore, is there a summary print option to get information of the current facet resolution? I am not sure, if the polydata generated by the Graphics display is the exact representation of the facets, if no mesh is displayed. Can someone confirm on this, if this is the case?

The only way I found for now to summarize the resolution of the facet meshes is by using the following lines of code:
```python
display = Graphics(model=model)
display()
number_of_cells = 0
for mesh in display.get_face_mesh_data():
    number_of_cells += mesh.poly_data.GetNumberOfCells()
print(f"number_of_cells = {number_of_cells}")
```
This is very slow, because I first have to wait for the display and cancel it in order to get access to the poly_data cells.

I try to import Rhino models and tried refaceting with ACIS and PARASOLID using COARSE, MEDIUM, FINE and CUSTOM. PARASOLID generates always facet meshes with at least one order less cells than ACIS. Has anybody an explanation for it? With PARASOLID it seems that I cannot change the facet resolution at all. 
I tried the following and the facet mesh stayed the same with PARASOLID:    
```python
import_params = prime.ImportCadParams(
        model=model,
        length_unit=prime.LengthUnit.MM,
        part_creation_type=prime.PartCreationType.MODEL,
        cad_reader_route=prime.CadReaderRoute.PROGRAMCONTROLLED,
        geometry_transfer=False, refacet=True,
        cad_refaceting_params=prime.CadRefacetingParams(model=model, cad_faceter=prime.CadFaceter.PARASOLID,
                                                        faceting_resolution=prime.CadRefacetingResolution.CUSTOM,
                                                       custom_surface_deviation_tolerance=0.1,
                                                       custom_normal_angle_tolerance=1))
```
Using ACIS produced a very fine facet mesh.

**Comment by dkowollik** [comment]
Hi everyone, 

Now, I tried a simplified model in different file formats with the two cad_faceter ACIS and PARASOLID. There must be a bug with the combination Rhino files and PARASOLID. Refaceting with *.x_t and PARASOLID works as expected. Refaceting with *.3dm and PARASOLID does not change anything. Therefore, I create a bug report for this issue and upload the simplified model there. 

Nevertheless, has anyone an answer to the faceting summary question I described above?

**Comment by waltersma** [comment, marked as answer]
Hi @dkowollik,

Cad refaceting is only available during import. 

The graphics display does represent the faceting if the topo faces are displayed without mesh.  

If you are only checking the facet count you should see this reported in the logs during import.

If you require more information you can convert the faceted geometry to a mesh and then use any of the summarisation routines on the faceting.  However, this does delete the topology and you would need to import the cad again to continue.  

For example:
```
for part in model.parts:
    part.delete_topo_entities(
        prime.DeleteTopoEntitiesParams(model, delete_geom_zonelets=False)
    )
    part._print_mesh = True
    print(part)
```

I confirm I can see an issue with the Parasolid faceting for rhino files.  I will update the issue accordingly.  

Thanks

**Reply by dkowollik** [reply]
Hi @waltersma,

Thank you for your answer. Where do I find these logs you mentioned above? 

The returned object of the import_cad method is  of type ImportCadResults. The object is completely empty, if no errors occur. It would be great to have the used parameters passed to this object. You are using a lot of default parameters. I can only find them, if I jump to the code within my IDE or debug into it. In addition to this, it would be great to get the infos of the refacet meshes within this object. A print_refacet_summary would be awesome. What do you think about this?

**Reply by waltersma** [reply]
Hi @dkowollik,

This information should be printed to the console during import and if you wish to attach a logger a basic example can be found [here](https://prime.docs.pyansys.com/version/stable/user_guide/logging.html).
